src/specular/stages/batches.rs

```diff
@@ -233,40 +233,49 @@ fn decode_batches_v0(
     for batch_list in batch_lists.iter() {
         // Decode the first l2 block number at offset 0.
         let batch_first_l2_num: u64 = batch_list.val_at(0)?;
-        // Check for duplicates.
-        if batch_first_l2_num < local_l2_num {
+        if batch_first_l2_num <= local_l2_num {
+            // We've already (partially) seen this batch.
             tracing::warn!(
                 "BatcherTx batches already seen | safe_head={} first_l2_block_num={}",
                 local_l2_num,
                 batch_first_l2_num
             );
-            eyre::bail!("BatcherTx batches already seen");
-        }
-        // Insert empty batches for missing blocks.
-        for i in local_l2_num + 1..batch_first_l2_num {
-            let batch = SpecularBatchV0 {
-                epoch_num,
-                epoch_hash,
-                timestamp: (i - safe_l2_num) * config.chain.blocktime + safe_l2_ts,
-                transactions: Vec::new(),
-                l2_block_number: i,
-                l1_inclusion_block: state.current_epoch_num,
-                l1_oracle_values: None,
-            };
-            tracing::trace!(
-                "inserting empty batch | num={} ts={}",
-                batch.l2_block_number,
-                batch.timestamp,
-            );
-            batches.push(batch);
+        } else {
+            // Insert empty batches for potential missing blocks.
+            for i in local_l2_num + 1..batch_first_l2_num {
+                let batch = SpecularBatchV0 {
+                    epoch_num,
+                    epoch_hash,
+                    timestamp: (i - safe_l2_num) * config.chain.blocktime + safe_l2_ts,
+                    transactions: Vec::new(),
+                    l2_block_number: i,
+                    l1_inclusion_block: state.current_epoch_num,
+                    l1_oracle_values: None,
+                };
+                tracing::info!(
+                    "inserting empty batch | num={} ts={}",
+                    batch.l2_block_number,
+                    batch.timestamp,
+                );
+                batches.push(batch);
+            }
+            // Update the local l2 block number.
+            // We're supposed to have inserted empty batches until right before the first batch in the list.
+            local_l2_num = batch_first_l2_num - 1;
         }
-        // Update the local l2 block number.
-        // We're supposed to have inserted empty batches until right before the first batch in the list.
-        local_l2_num = batch_first_l2_num - 1;
         let batch_first_l2_ts =
             (batch_first_l2_num - safe_l2_num) * config.chain.blocktime + safe_l2_ts;
         // Decode the transaction batches at offset 1.
         for (batch, idx) in batch_list.at(1)?.iter().zip(0u64..) {
+            if batch_first_l2_num + idx <= local_l2_num {
```

Comment:
nit since we use `batch_first_l2_num + idx` in 3 lines
```suggestion
            let l2_block_number = batch_first_l2_num + idx;
            if l2_block_number <= local_l2_num {
```